backend/src/user/auth.py

```diff
@@ -0,0 +1,209 @@
+import datetime as dt
+from typing import Dict, List, Optional
+
+from fastapi import Depends, FastAPI, HTTPException, Request, Response, status
+from fastapi.openapi.models import OAuthFlows as OAuthFlowsModel
+from fastapi.responses import HTMLResponse, RedirectResponse
+from fastapi.security import OAuth2, OAuth2PasswordRequestForm
+from fastapi.security.utils import get_authorization_scheme_param
+from fastapi.templating import Jinja2Templates
+from jose import JWTError, jwt
+from passlib.handlers.sha2_crypt import sha512_crypt as crypto
+from pydantic import BaseModel
+from rich import inspect, print
+from rich.console import Console
+
+console = Console()
+
+router = APIRouter()
+
+# --------------------------------------------------------------------------
+# Authentication logic
+# --------------------------------------------------------------------------
+class OAuth2PasswordBearerWithCookie(OAuth2):
+    """
+    This class is taken directly from FastAPI:
+    https://github.com/tiangolo/fastapi/blob/26f725d259c5dbe3654f221e608b14412c6b40da/fastapi/security/oauth2.py#L140-L171
+    
+    The only change made is that authentication is taken from a cookie
```

Comment:
What is the advantage of using the cookie instead of the header for the authentication?